### DISCUSSION TITLE
Is it possible to add videos in bulk by a script instead of using the upload page of MediaCMS?

### DISCUSSION THREAD

**Discussion by foeyonghai** [mediacms-io/mediacms#960] [General]
Hello all,

I have a lot of videos that I want to add without using the upload feature of MediaCMS because I prefer to script this when possible. Maybe someone already did this before and wants to share his/her experiences? The reason that I want to script this is because the upload page of MediaCMS often gets erros like "upload failed", where afterwards strangely enough the upload succeeded. This is just one example. Maybe the API can be used for example?

I have limited experience in Python but I am learning and also really want to learn.

Many thanks for your thoughts!

**Comment by skydudie** [comment]
Using the Django shell you can run this script that will allow you to import all the files within a directory without uploading and having a copy created.

Enter the shell
`source  /home/mediacms.io/bin/activate`
`python manage.py shell`

Then run the script below, remember to swap out the directory_path to the directory that you want to import.

```
import os
from django.utils import timezone
from myapp.models import Media

# Change to your own directory that contains the files you want to import
directory_path = '/home/mediacms.io/mediacms/media_files/original/user/admin'

# Generate a list of all files/directories within the folder.
file_list = os.listdir(directory_path)

# Initialize a counter
counter = 1

# Loop through the list
for index, filename in enumerate(file_list, start=1):
    file_path = os.path.join(directory_path, filename)

    # Check if the path is a regular file (not a directory)
    if os.path.isfile(file_path):
        # Check if a Media object with this media_file already exists
        if not Media.objects.filter(media_file=file_path).exists():
            # Create a new Media object
            media = Media(
                user_id=1,
                media_file=file_path,
            )

            # Save the Media object
            media.save()

            # Print a message with the counter
            print(f"Processed file {counter}: {file_path}")

            # Increment the counter
            counter += 1
        else:
            print(f"Media file '{file_path}' already exists in the database.")
    else:
        print(f"Skipping directory '{file_path}'.")


```

**Reply by Arsinine** [reply]
Hi @skydudie

I've ran this and it says all my files are processed, but im not seeing anything in the mediacms UI even after restarting the container.
I can see under manage media that all of the imports have failed but theres no way of re-processing the videos.
![image](https://github.com/user-attachments/assets/5e1d9b35-f1be-48a1-b379-bcefa94b664e)

**Reply by foeyonghai** [reply]
Hi @fluxtheory, could you try it again after changing DO_NOT_TRANSCODE_VIDEO from False to True? 
You can find it in your settings.py on line 502, example: https://github.com/mediacms-io/mediacms/blob/main/cms/settings.py. Restarting your container is mandatory for this.

**Reply by Arsinine** [reply]
I nuked the volumes and containers and started over but no change unfortunately 
![image](https://github.com/user-attachments/assets/dcff8421-6460-4754-87dd-94a4aaf6f346)


![image](https://github.com/user-attachments/assets/686d5dce-5e74-4dba-a5ac-143f036e019e)

Videos are still unplayable.
![image](https://github.com/user-attachments/assets/8bf52a80-8c52-46f3-9023-1a7928ca4773)

Seems like its a 404 not found issue, I think I know what the culprit is. Seems like the spaces in the filenames is breaking the url

**Reply by skydudie** [reply]
Hi @fluxtheory, this looks like it might be a permissions issue. Might want to check that your files are accessible by the webserver.

**Comment by shootie22** [comment]
hey, I'm trying to use the above script, but I am wondering if this is a typo, since it doesn't work:
`from myapp.models import Media`
What should "myapp" be set to? I tried setting it to mediacms, django, but it doesn't seem to work, erroring with "ModuleNotFoundError: No module named 'myapp' "

--- edit ---

through some trial & error I managed to get it to run. steps:

make sure to:
- replace `from myapp.models import Media` with `from users.models import Media` - _myapp_ to _users_
- the videos need to be under `/home/mediacms.io/mediacms/media_files/YOUR_VID_FOLDER/`, otherwise the script will throw a SuspiciousFileOperation. if you have them on another server, you can mount them into a folder using sshfs.

steps:
1. store the script somewhere
2. `source /home/mediacms.io/bin/activate`
3. `cd /home/mediacms.io/mediacms`
4. `sudo ../bin/python manage.py shell` - running without sudo gives me a PermissionDenied error
5. in the python shell, use `exec(open('/path/to/your/script.py').read())` to run the script

this feels very roundabout, so there must be a proper way to do it, but this worked for me.

**Reply by shootie22** [reply]
I've been at it for the past couple of days, and the main issue that I am running into right now is that while it works pretty well, imported .mkv videos show up as a black screen when trying to play them. They generate thumbnails fine, but as pointed out in #976, they don't play once you have `DO_NOT_TRANSCODE_VIDEO = True` in `settings.py`.  

By checking the console while hitting refresh on a page displaying an .mkv video, I can see that it says that the video file "doesn't exist". That turns out to be coming from either:
- around line 5000:
```
if (1 === a.length && "running" === a[0].encodings_status && (s = "encodingRunning", c = "Media encoding is currently running. Try again in few minutes."), null !== s) switch (s) {
                                case "encodingRunning":
                                case "encodingPending":
                                case "encodingFailed":
                                    break;
                                default:
                                    f.warn("VIDEO DEBUG:", "Video files don't exist")
                            }
```
- or around line 460:
```
else switch (o.MediaPageStore.get("media-load-error-type")) {
                            case "encodingRunning":
                            case "encodingPending":
                            case "encodingFailed":
                                break;
                            default:
                                O.warn("VIDEO DEBUG:", "Video files don't exist")
                        }
```

Will try to poke around a bit more, maybe I will make it work. But for now, an alternative would be to convert all my .mkv files into .mp4 files, which I'm not the most excited by as I have thousands of them.
